### PR TITLE
Db::numRows is unreliable and Db::Affected_Rows throws exception if previous query failed silently

### LINKED ISSUE #25515: The method Db::numRows() is unreliable
#### Describe the bug

I used the method `Db::numRows()` instead of `Db:Affected_Rows()` (with `DBO` and Db Caching enabled) already since a long time. A just got bitten by a bug difficult to reproduce.
This is when I noticed, that `Db::numRows()` isn't used anymore in the whole codebase, and that in fact it cannot work reliably.
- It is meant to be used with `executeS` and `getRow`, but is not reliable for this with DBO
- With DBO it works with insert, update and delete queries, but the caching logic is wrong for those and thus is not reliable neither.

#### Steps to Reproduce

**Screenshots**

#### Additional information

* PrestaShop version: any
* PHP version: PHP 7.x

### PR DESCRIPTION
**Pull request by lmeyer1**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Improvement of phpdoc, deprecation of unused and unreliable method Db::numRows() 
| Type?             | documentation
| Category?         | CO
| BC breaks?        | no The deprecated function is not used in the whole codebase 
| Deprecations?     | yes Db::numRows()
| Fixed ticket?     | Fixes #25515.
| How to test?      | 
| Possible impacts? | Should not have impacts, since the method is not used anymore.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25514)

### CODE REVIEW

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

File: classes/db/Db.php [hunk 1]
```diff
@@ -704,6 +706,11 @@ public function getValue($sql, $use_cache = true)
 
     /**
      * Get number of rows for last result.
+     * Must be used solely with select queries.
+     * Results are unreliable with DbPDO (https://www.php.net/manual/en/pdostatement.rowcount.php)
+     * With Db Caching enabled, it works only after `executeS()` and `getRow()`
+     *
+     * @deprecated Never used throughout current codebase
```
Comment (eternoendless) [inline classes/db/Db.php:713]: It is used:

Screenshot 2021-08-20 at 14 38 31

File: classes/db/DbPDO.php [hunk 2]
```diff
@@ -240,6 +240,8 @@ public function Insert_ID()
      *
      * @see DbCore::Affected_Rows()
      *
+     * @throws Error if previous query failed silently (`$this->result` is not a `PDOStatement`)
+     *
```
Comment (eternoendless) [inline classes/db/DbPDO.php:244]: It would be best to throw an Exception if `$this->result` is not what is expected

File: classes/db/Db.php (hunk 1 above)
Comment (lmeyer1) [reply classes/db/Db.php:713]: @eternoendless
I don't know how I searched then. I think best is to close the PR and the issue.

**Comment by lmeyer1** [comment]
There is an real underlying problem with db caching and this method, but the way to solve it is wrong. Deprecation is not an option, since the method is used several times
Also, I was not able to write a test case to reproduce the issue, the cache issue is too intricate and random.

**Pull request closed without merging**